pkg/databrickssql/connector.go

```diff
@@ -0,0 +1,169 @@
+
+		temporaryCredentials: credentials.NewCredentials(NewTemporaryCredentialsProvider(srcCredentials, awsRegion)),
+
+		storageURL: storageURL,
+		columns:    nil,
+	}, nil
+}
+
+func (dc *DatabricksConnector) InitSchema(columns []cloudstorage.TableCol) error {
+	if len(dc.columns) != 0 {
+		return nil
+	}
+	if len(columns) == 0 {
+		return errors.New("Columns in schema is empty")
+	}
+	dc.columns = columns
+	log.Info("table columns initialized", zap.Any("Columns", columns))
+	return nil
+}
+
+func (dc *DatabricksConnector) CopyTableSchema(sourceDatabase string, sourceTable string, sourceTiDBConn *sql.DB) error {
+	dropTableSQL := GenDropTableSQL(sourceTable)
+	_, err := dc.db.Exec(dropTableSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = dc.setColumns(sourceDatabase, sourceTable, sourceTiDBConn)
+	createTableSQL, err := GenCreateTableSQL(sourceTable, dc.columns)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	log.Info("Creating table in Databricks Warehouse", zap.String("query", createTableSQL))
+
+	_, err = dc.db.Exec(createTableSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Info("Successfully copying table scheme", zap.String("database", sourceDatabase), zap.String("table", sourceTable))
+	return nil
+}
+
+func (dc *DatabricksConnector) LoadSnapshot(targetTable, filePrefix string, onSnapshotLoadProgress func(loadedRows int64)) error {
+	if err := LoadCSVFromS3(dc.db, dc.columns, targetTable, dc.storageURL, filePrefix, dc.temporaryCredentials); err != nil {
+		return errors.Trace(err)
+	}
+	log.Info("Successfully load snapshot", zap.String("table", targetTable), zap.String("filePrefix", filePrefix))
+	return nil
+}
+
+func (dc *DatabricksConnector) ExecDDL(tableDef cloudstorage.TableDefinition) error {
+	if len(dc.columns) == 0 {
+		return errors.New("Columns not initialized. Maybe you execute a DDL before all DMLs, which is not supported now.")
+	}
+	ddls, err := GenDDLViaColumnsDiff(dc.columns, tableDef)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(ddls) == 0 {
+		log.Info("No need to execute this DDL in Databricks", zap.String("ddl", tableDef.Query))
+		return nil
+	}
+	// One DDL may be rewritten to multiple DDLs
+	for _, ddl := range ddls {
+		_, err := dc.db.Exec(ddl)
+		if err != nil {
+			log.Error("Failed to executed DDL", zap.String("received", tableDef.Query), zap.String("rewritten", strings.Join(ddls, "\n")))
+			return errors.Annotate(err, fmt.Sprint("failed to execute", ddl))
+		}
+	}
+	// update columns
+	dc.columns = tableDef.Columns
+	log.Info("Successfully executed DDL", zap.String("received", tableDef.Query), zap.String("rewritten", strings.Join(ddls, "\n")))
+	return nil
+}
+
+func (dc *DatabricksConnector) LoadIncrement(tableDef cloudstorage.TableDefinition, uri *url.URL, filePath string) error {
+	absolutePath := fmt.Sprintf("%s://%s%s/%s", uri.Scheme, uri.Host, uri.Path, filePath)
+	incrTableColumns := bigquerysql.GenIncrementTableColumns(tableDef.Columns)
```

Comment:
it may be better to move `GenIncrementTableColumns` to `utils/`?

Reply:
Sure